tools/closed-by-commits

```diff
@@ -0,0 +1,361 @@
+            urls.add(url)
+
+        return frozenset(urls)
+
+    except requests.exceptions.RequestException as e:
+        assert e.response
+        if e.response.status_code != 429:
+            raise
+        retry_after = int(e.response.headers["Retry-After"]) + 1
+        time.sleep(retry_after)
+        return search_czo_for_number(number)
+
+
+@dataclass
+class Issue:
+    number: int
+    title: str
+    czo_urls: set[str] = field(default_factory=set)
+    closed_by_prs: list["PullRequest"] = field(default_factory=list)
+
+    @override
+    def __hash__(self) -> int:
+        return hash(self.number)
+
+    @override
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, Issue) and self.number == other.number
+
+
+@dataclass
+class PullRequest:
+    number: int
+    title: str
+    czo_urls: set[str] = field(default_factory=set)
+
+    @override
+    def __hash__(self) -> int:
+        return hash(self.number)
+
+    @override
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, PullRequest) and self.number == other.number
+
+
+class CommitRangeAnalyzer:
+    COMMIT_PRS_QUERY = """
+    query($oid: GitObjectID!) {
+      repository(owner: "zulip", name: "zulip") {
+        object(oid: $oid) {
+          ... on Commit {
+            messageBody
+            associatedPullRequests(first: 10) {
+              nodes {
+                number
+                title
+                url
+                body
+                comments(first: 100) {
+                  nodes {
+                    body
+                  }
+                }
+                closingIssuesReferences(first: 50) {
+                  nodes {
+                    number
+                    title
+                    url
+                    body
+                    comments(first: 100) {
+                      nodes {
+                        body
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    """
+
+    ISSUE_QUERY = """
+    query($number: Int!) {
+      repository(owner: "zulip", name: "zulip") {
+        issue(number: $number) {
+          number
+          title
+          body
+          comments(first: 100) {
+            nodes {
+              body
+            }
+          }
+        }
+      }
+    }
+    """
+
+    def __init__(self, token: str) -> None:
+        self.github = Github(auth=Auth.Token(token))
+
+    @staticmethod
+    def _extract_czo_urls(text: str | None) -> set[str]:
+        if not text:
+            return set()
+
+        matches = re.findall(r"https://chat\.zulip\.org/[^\s\)\]\>]+", text)
+
+        urls = set()
+        for url in matches:
+            if "/topic/" not in url:
+                continue
+            urls.add(re.sub(r"/(with|near)/\d+$", "", url))
```

Comment:
What's your thinking behind stripping off /with/… and /near/…?

One effect of doing that is that if the topic has been resolved or otherwise moved, then the link in this output won't lead to the same place the original link would (and typically would lead to an empty conversation).